src/suggestion.c

```diff
@@ -0,0 +1,399 @@
+
+                // String already exists in suggestion set
+
+                // If we found the string in less edits, update its score
+                if (set[i]->edits_left < edits_left) {
+                    set[i]->edits_left = edits_left;
+                }
+
+                // We don't want to trigger adding in place of the worst match
+                i = max_matches + 1;
+
+                break;
+            }
+        }
+
+        // If we hit the maximum amount of items, then see if we can add it in place of the worst match
+        if (i == max_matches) {
+
+            i--;
+
+            // Sort the set by edits_left and break ties alphabetically
+            qsort(set, max_matches, sizeof(match_t*), cmp_match);
+
+            // If we found it in less edits or match was alphabetically first
+            if (set[i]->edits_left < edits_left 
+                || (set[i]->edits_left == edits_left && (strncmp(set[i]->str, s, MAXLEN) > 0))) {
+
+                set[i]->str = malloc(sizeof(char) * MAXLEN);
+                if (set[i]->str == NULL) {
+                    return EXIT_FAILURE;
+                }
+                strcpy(set[i]->str, s);
+                set[i]->edits_left = edits_left;
+
+                if (set[i]->str == NULL) {
+                    // strcpy failed
+                    return EXIT_FAILURE;
+                }
+            }
+        }
+    }
+
+    return EXIT_SUCCESS;
+}
+
+// Look at suggestion.h for documentation
+int suggestions(match_t **set, trie_t *t, char *prefix, char *suffix, int edits_left, int max_matches) {
+
+    char* s;
+    int rc = 0;
+    
+    // Since prefix and suffix can have max length MAXLEN
+    s = malloc(sizeof(char) * (MAXLEN + 1) * 2);
+    if (s == NULL) {
+        return rc + EXIT_FAILURE;
+    }
+
+    // Now put prefix and suffix together
+    strncpy(s, prefix, MAXLEN);
+    strncat(s, suffix, MAXLEN);
+
+    if (attempt_add(set, t, s, edits_left, max_matches) != EXIT_SUCCESS)  {
+        return EXIT_FAILURE;
+    }
+
+    if (edits_left <= 0) {
+        // Hooray for exit conditions!
+
+        free(s);
+        return rc;
+    }
+
+    // Make sure we aren't at the end of the suffix
+    if (suffix[0] != '\0') {
+
+        rc += move_on(set, t, prefix, suffix, edits_left, max_matches);
+
+        rc += try_delete(set, t, prefix, suffix, edits_left, max_matches);
+
+        rc += try_replace(set, t, prefix, suffix, edits_left, max_matches);
+
+        rc += try_swap(set, t, prefix, suffix, edits_left, max_matches);
+    }
+
+    // This one doesn't need any fancy suffix stuff
+    rc += try_insert(set, t, prefix, suffix, edits_left, max_matches);
+
+    free(s);
+
+    return rc;
+}    
+
+match_t** suggestion_set_new(trie_t *t, char *str, int max_edits, int max_matches) {
+
+    assert(t != NULL);
+    assert(str != NULL);
+
+    // We'll allocate 4 times as many matches so we can make sure we get the closest matches
```

Comment:
What do you mean? Aren't you just callocing the max number of matches?

Reply:
@dwahme

Reply:
Good catch, this was an out of date comment from a previous version of the implementation. This version of the implementation is more elegant and safer, so 4x was unnecessary. Removed.